webgoat-lessons/sql-injection/src/main/java/org/owasp/webgoat/sql_injection/introduction/SqlInjectionLesson8.java

```diff
@@ -0,0 +1,142 @@
+
+        try (Connection connection = dataSource.getConnection()) {
+            try {
+                Statement statement = connection.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_UPDATABLE);
+                log(connection, query);
+                ResultSet results = statement.executeQuery(query);
+
+                if (results.getStatement() != null) {
+                    if (results.first()) {
+                        output.append(generateTable(results));
+                        results.last();
+
+                        if (results.getRow() > 1) {
+                            // more than one record, the user succeeded
+                            return success(this).feedback("sql-injection.8.success").output(output.toString()).build();
+                        } else {
+                            // only one record
+                            return failed(this).feedback("sql-injection.8.one").output(output.toString()).build();
+                        }
+
+                    } else {
+                        // no results
+                        return failed(this).feedback("sql-injection.8.no.results").build();
+                    }
+                } else {
+                    return failed(this).feedback("sql-injection.error").build();
+                }
+            } catch (SQLException e) {
+                return failed(this).feedback("sql-injection.error").output("<br><span class='feedback-negative'>" + e.getMessage() + "</span>").build();
+            }
+
+        } catch (Exception e) {
+            return failed(this).feedback("sql-injection.error").output("<br><span class='feedback-negative'>" + e.getMessage() + "</span>").build();
+        }
+    }
+
+    public static String generateTable(ResultSet results) throws SQLException {
+        ResultSetMetaData resultsMetaData = results.getMetaData();
+        int numColumns = resultsMetaData.getColumnCount();
+        results.beforeFirst();
+        StringBuffer table = new StringBuffer();
+        table.append("<table>");
+
+        if (results.next()) {
+            table.append("<tr>");
+            for (int i = 1; i < (numColumns + 1); i++) {
+                table.append("<th>" + resultsMetaData.getColumnName(i) + "</th>");
+            }
+            table.append("</tr>");
+
+            results.beforeFirst();
+            while (results.next()) {
+                table.append("<tr>");
+                for (int i = 1; i < (numColumns + 1); i++) {
+                    table.append("<td>" + results.getString(i) + "</td>");
+                }
+                table.append("</tr>");
+            }
+
+        } else {
+            table.append("Query Successful; however no data was returned from this query.");
+        }
+
+        table.append("</table>");
+        return (table.toString());
+    }
+
+    public static void log(Connection connection, String action) {
+        action = action.replace('\'', '"');
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;action&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;connection&#39;. The same parameter type is validated here for example: webgoat-container/src/main/java/org/owasp/webgoat/AjaxAuthenticationEntryPoint.java:48. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, reduce the method&#39;s visibility.

Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).